lib/parser/dockerfile/add_copy.go

```diff
@@ -14,30 +14,59 @@
 
 package dockerfile
 
-import "strings"
+import (
+	"fmt"
+	"strings"
+)
 
 type addCopyDirective struct {
 	*baseDirective
-	Chown string
-	Srcs  []string
-	Dst   string
+	Chown         string
+	PreserveOwner bool
+	Srcs          []string
+	Dst           string
 }
 
 // Variables:
 //   Replaced from ARGs and ENVs from within our stage.
 // Formats:
+//   ADD/COPY [--archive] <src>... <dest>
+//   ADD/COPY [--archive] ["<src>",... "<dest>"]
 //   ADD/COPY [--chown=<user>:<group>] ["<src>",... "<dest>"]
 //   ADD/COPY [--chown=<user>:<group>] <src>... <dest>
 func newAddCopyDirective(base *baseDirective, args []string) (*addCopyDirective, error) {
 	if len(args) == 0 {
 		return nil, base.err(errMissingArgs)
 	}
 
+	// Check the flag numbers here since we only allow zero or one flag here.
+	var chownCount, archiveCount int
 	var chown string
-	if val, ok, err := parseFlag(args[0], "chown"); err != nil {
-		return nil, base.err(err)
-	} else if ok {
-		chown = val
+	var preserveOwner bool
+	for _, arg := range args[:len(args)-1] {
+		if strings.HasPrefix(arg, "--chown") {
+			if val, ok, err := parseStringFlag(arg, "chown"); err != nil {
+				return nil, base.err(err)
+			} else if ok {
+				chown = val
+				chownCount++
+				continue
+			}
+		}
+
+		if strings.HasPrefix(arg, "--archive") {
+			if err := parseBoolFlag(arg, "archive"); err == nil {
+				archiveCount++
+				preserveOwner = true
+			} else {
+				return nil, fmt.Errorf("archive flag format is wrong")
+			}
+		}
+	}
+
+	if archiveCount+chownCount >= 2 {
+		return nil, base.err(fmt.Errorf("arguments shouldn't contain more than one flag [--chown or --archive]"))
```

Comment:
uppercase `argument`?

Reply:
it should be `argument` here and I think we should keep it lowercase.

1.   keep consistent with `fmt.Errof` elsewhere.
2.   isn't it golang style to keep it lowercase?